File: android/src/main/kotlin/tools/isekai/terminal/ui/TerminalHyperlink.kt
package tools.isekai.terminal.ui

import uniffi.isekai_terminal_core.ScreenUpdate

/**
 * OSC 8(タスク#40)ハイパーリンクのタップUI(タスク#52)向けの純粋ロジック。
 *
 * hit-test自体(画面座標→セル→リンクURL)は「UI表示だけに閉じた判断」であり
 * rust-ssot原則([.claude/rules/rust-ssot.md])の対象外——Rust側は既に
 * `CellData.linkId`/`ScreenUpdate.linkTable`としてintern済みのURL文字列を
 * 公開済みで、ここではその配列を読むだけで新しい状態機械は作らない。
 *
 * スキーム許可判定([isOpenableHyperlinkScheme])はセキュリティポリシーであり、
 * [RemoteClipboardPolicy]と同じ理由(接続先ホストの出力=信頼できない入力)で
 * 独立した純粋関数として切り出しAndroidフレームワーク非依存でテストする。
 * 呼び出し元(`TerminalScreen.kt`)は、タップされたセルにリンクがあり、かつ
 * スキームが許可リストに含まれる場合のみURL全文の確認ダイアログを表示し、
 * ユーザーが確認した場合のみ`ACTION_VIEW` Intentを発行する。
 */

/**
 * [update]の([row], [col])セル(0-indexed)が指すハイパーリンクURLを返す。
 * リンクなし・範囲外・`linkId`が`linkTable`の範囲外(本来起こらないはずだが、
 * 呼び出し側からの防御的な扱いとしてクラッシュではなくnullを返す)の場合はnull。
 */
fun linkUrlAtCell(update: ScreenUpdate, row: Int, col: Int): String? {
    val cols = update.cols.toInt()
    val rows = update.rows.toInt()
    if (cols <= 0 || rows <= 0 || row !in 0 until rows || col !in 0 until cols) return null
    val cells = update.cells
    val index = row * cols + col
    if (index !in cells.indices) return null
    val linkId = cells[index].linkId ?: return null
    return update.linkTable.getOrNull(linkId.toInt())
}

/** URL文字列の先頭にある `scheme:` を抽出する(RFC 3986 3.1 の生成規則に準拠)。 */
private val SCHEME_REGEX = Regex("^([A-Za-z][A-Za-z0-9+.-]*):")

/**
 * リモート(信頼できないホスト出力)由来のURLを、タップ時に`ACTION_VIEW`へ渡して
 * 安全に開いてよいかどうか。`http`/`https`のみ許可する——`intent://`は任意のAndroid
 * コンポーネント起動、`file://`はローカルファイル露出、`javascript:`はWebView文脈での
 * スクリプト実行に悪用されうるため、無条件で`ACTION_VIEW`へ渡さない
 * (タスク#52 Fableレビュー2次のセキュリティ要件)。
 */
fun isOpenableHyperlinkScheme(url: String): Boolean {
    val scheme = SCHEME_REGEX.find(url)?.groupValues?.get(1)?.lowercase() ?: return false
    return scheme == "http" || scheme == "https"
}

File: ios/Sources/IsekaiTerminalCoreLogic/TerminalHyperlink.swift
import Foundation

/// OSC 8(タスク#40)ハイパーリンクのタップUI(タスク#52)向けの純粋ロジック。
/// Android版`TerminalHyperlink.kt`と対称。
///
/// hit-test自体(画面座標→セル→リンクURL)は「UI表示だけに閉じた判断」であり
/// rust-ssot原則(`.claude/rules/rust-ssot.md`)の対象外——Rust側は既に
/// `CellData.linkId`/`ScreenUpdate.linkTable`としてintern済みのURL文字列を
/// 公開済みで、ここではその配列を読むだけで新しい状態機械は作らない。
///
/// スキーム許可判定(`isOpenableHyperlinkScheme`)はセキュリティポリシーであり、
/// 接続先ホストの出力(=信頼できない入力)を無条件で`UIApplication.open`へ渡さない
/// ための独立した純粋関数(タスク#52 Fableレビュー2次の要件、Android版
/// `RemoteClipboardPolicy`と同じ考え方)。呼び出し元(`TerminalScreenView.swift`)は、
/// タップされたセルにリンクがあり、かつスキームが許可リストに含まれる場合のみ
/// URL全文の確認ダイアログを表示し、ユーザーが確認した場合のみ`UIApplication.open`を呼ぶ。

/// `update`の(`row`, `col`)セル(0-indexed)が指すハイパーリンクURLを返す。
/// リンクなし・範囲外・`linkId`が`linkTable`の範囲外(本来起こらないはずだが、
/// 呼び出し側からの防御的な扱いとしてクラッシュではなくnilを返す)の場合はnil。
public func linkURL(at update: ScreenUpdate, row: Int, col: Int) -> String? {
    let cols = Int(update.cols)
    let rows = Int(update.rows)
    guard cols > 0, rows > 0, row >= 0, row < rows, col >= 0, col < cols else { return nil }
    let index = row * cols + col
    guard index >= 0, index < update.cells.count else { return nil }
    guard let linkId = update.cells[index].linkId else { return nil }
    let tableIndex = Int(linkId)
    guard tableIndex >= 0, tableIndex < update.linkTable.count else { return nil }
    return update.linkTable[tableIndex]
}

/// URL文字列の先頭にある`scheme:`を抽出する(RFC 3986 3.1の生成規則に準拠)。
private let hyperlinkSchemePattern = try? NSRegularExpression(pattern: "^([A-Za-z][A-Za-z0-9+.-]*):")

/// リモート(信頼できないホスト出力)由来のURLを、タップ時に`UIApplication.open`へ
/// 渡して安全に開いてよいかどうか。`http`/`https`のみ許可する——`intent://`相当の
/// 任意コンポーネント起動や`file://`によるローカルファイル露出、`javascript:`の
/// スクリプト実行に悪用されうるため、無条件で開かない
/// (タスク#52 Fableレビュー2次のセキュリティ要件)。
public func isOpenableHyperlinkScheme(_ url: String) -> Bool {
    guard let regex = hyperlinkSchemePattern else { return false }
    let range = NSRange(url.startIndex..<url.endIndex, in: url)
    guard let match = regex.firstMatch(in: url, range: range),
          let schemeRange = Range(match.range(at: 1), in: url) else { return false }
    let scheme = url[schemeRange].lowercased()
    return scheme == "http" || scheme == "https"
}
